### ISSUE TITLE
How to convert a build"model" to just "model" class?

### ISSUE THREAD

**Issue by guilhermeparreira** [cvoeten/buildmer#8]
Hi There! This is not an issue. After variable selection via any `build*model` function, I want to do some diagnostics or contrast comparison. However, for doing that, it is usually required to use the class of the model, not the `buildmer` class. Is there a workaround for this?

Example:

```
dmaster <- structure(list(TempodeProva = c(778.433333333333, 798.816666666667, 
802.416666666667, 898.583333333333, 898.583333333333, 913.733333333333, 
1002.25, 1084.91666666667, 1094.91666666667, 1132.55, 1141.48333333333, 
1141.5, 1190.41666666667, 1210.51666666667, 1217.46666666667, 
1219.71666666667, 1230.6, 1248.95, 1260.51666666667, 1283.46666666667, 
1283.83333333333, 1309.16666666667), Ca = c(-0.28, -0.18, 0.36, 
-0.21, -0.53, -0.03, -0.0899999999999999, -0.11, -0.1, -0.11, 
-0.0700000000000001, 0.01, -0.05, -0.13, -0.1, -0.0700000000000001, 
-0.0800000000000001, -0.12, -0.12, -0.0700000000000001, 0.02, 
-0.12), Wtr = c(-0.399999999999999, -1.1, -0.200000000000003, 
2.1, -1.4, -0.800000000000004, -47.3, -32.1, -1.8, -39.2, -1.90000000000001, 
-1.3, 1.5, 1.1, -1.7, 1.9, 0.800000000000004, -3, 3.1, -0.399999999999999, 
-1.40000000000001, 0.100000000000001)), class = "data.frame", row.names = c(NA, 
-22L))
m2 <- buildgam(TempodeProva ~ 1 + s(Ca) + s(Wtr), 
           direction = "forward",
           data = dmasterm, 
           quiet = T,
           family = Gamma(link = "log"))
class(m2)
[1] "buildmer"
attr(,"package")
[1] "buildmer"
gam.check(m2)
Error: $ operator not defined for this S4 class
```
This is an weird error. But when I tried using `emmeans` or `DHARMa` packages, they don't recognized the model (it is not a supported model)..

```
m2 <- gam(TempodeProva ~ 1 + s(Ca) + s(Wtr), 
          data = dmasterm, 
          family = Gamma(link = "log"))
class(m2)
gam.check(m2)
# Works
```

My actual work around is using `formula` from the fit. But is there a better way of doing it?

Thanks!!

**Comment by cvoeten** [comment]
Yes, there is a better way! The actual model object is in the 'model' slot, i.e. `m2@model` gives you the raw fitted model object.

For most S3/S4 generics, buildmer exports wrapper functions, e.g. `logLik.buildmer <- function (object,...) logLik(object=object@model,...)`. So things like `logLik`, `coef`, `drop1`, etc should all work, being forwarded to the object in the 'model' slot. However, for technical reasons the package does not do this for generics that are defined by other packages (such as emmeans and DHARMa), and of course this can never be handled automatically for functions that are not generics but only accept their own internal objects (such as gam.check). In those cases, you need to explicitly take the 'model' slot yourself.

It is possible to have the buildmer package run some code when packages like emmeans are attached, to install its forwarding code onto the new generics that are exported by the newly-attached package. I recall emmeans itself doing something like this. I tried looking into this myself, but I couldn't figure out how to get it right. As they say, patches welcome! I guess this TODO is a reason to keep this report open after all, in case somebody else stumbles upon it and is more experienced with S3/S4 programming...
